src/lib/kdf/sp800_56a/sp800_56c_one_step.cpp

```diff
@@ -0,0 +1,192 @@
+      reset_h_callback(&h);
+
+      // 6.2 Compute K(i) = H(counter || Z || FixedInfo).
+      h.update_be(counter);
+      h.update(z);
+      h.update(fixed_info);
+      auto k_i = h.final();
+
+      // 6.3. Set Result(i) = Result(i−1) || K(i).
+      result.insert(result.end(), k_i.begin(), k_i.end());
+   }
+
+   // 7. Set DerivedKeyingMaterial equal to the leftmost L bits of Result(reps).
+   copy_mem(output_buffer, std::span(result).subspan(0, output_buffer.size()));
+}
+
+}  // namespace
+
+void SP800_56C_One_Step_Hash::kdf(uint8_t key[],
+                                  size_t key_len,
+                                  const uint8_t secret[],
+                                  size_t secret_len,
+                                  const uint8_t salt[],
+                                  size_t salt_len,
+                                  const uint8_t label[],
+                                  size_t label_len) const {
+   BOTAN_UNUSED(salt);
+   BOTAN_ARG_CHECK(salt_len == 0, "SP800_56A_Hash does not support a non-empty salt");
+
+   kdm_internal({key, key_len}, {secret, secret_len}, {label, label_len}, *m_hash, [](Buffered_Computation* kdf) {
+      HashFunction* hash = dynamic_cast<HashFunction*>(kdf);
+      BOTAN_ASSERT_NONNULL(hash);
+      hash->clear();
+   });
+}
+
+std::string SP800_56C_One_Step_Hash::name() const {
+   return fmt("SP800-56A({})", m_hash->name());
+}
+
+std::unique_ptr<KDF> SP800_56C_One_Step_Hash::new_object() const {
+   return std::make_unique<SP800_56C_One_Step_Hash>(m_hash->new_object());
+}
+
+SP800_56C_One_Step_HMAC::SP800_56C_One_Step_HMAC(std::unique_ptr<MessageAuthenticationCode> mac) :
+      m_mac(std::move(mac)) {
+   // TODO: we need a MessageAuthenticationCode::is_hmac
+   if(!m_mac->name().starts_with("HMAC(")) {
+      throw Algorithm_Not_Found("Only HMAC can be used with SP800_56A_HMAC");
+   }
+}
+
+void SP800_56C_One_Step_HMAC::kdf(uint8_t key[],
+                                  size_t key_len,
+                                  const uint8_t secret[],
+                                  size_t secret_len,
+                                  const uint8_t salt[],
+                                  size_t salt_len,
+                                  const uint8_t label[],
+                                  size_t label_len) const {
+   kdm_internal({key, key_len}, {secret, secret_len}, {label, label_len}, *m_mac, [&](Buffered_Computation* kdf) {
+      MessageAuthenticationCode* kdf_mac = dynamic_cast<MessageAuthenticationCode*>(kdf);
+      BOTAN_ASSERT_NONNULL(kdf_mac);
+      kdf_mac->clear();
+      // 4.1 Option 2 and 3 - An implementation dependent byte string, salt,
+      //     whose (non-null) value may be optionally provided in
+      //     OtherInput, serves as the HMAC#/KMAC# key ..
+
+      // SP 800-56A specifies if the salt is empty then a block of zeros
+      // equal to the hash's underlying block size are used. However this
```

Comment:
This is an issue in the original comment but we should fix it while we're here

> However *for HMAC* this is equivalent

Since that's not generally true for eg KMAC or whatever